doubly_linked_list/doubly_linked_list.py

```diff
@@ -1,68 +1,161 @@
 
-  """Rearranges this ListNode's previous and next pointers
+    def insert_before(self, value):
+        current_prev = self.prev
+        self.prev = ListNode(value, current_prev, self)
+        if current_prev:
+            current_prev.next = self.prev
+
+    """Rearranges this ListNode's previous and next pointers
   accordingly, effectively deleting this ListNode."""
-  def delete(self):
-    if self.prev:
-      self.prev.next = self.next
-    if self.next:
-      self.next.prev = self.prev
+
+    def delete(self):
+        if self.prev:
+            self.prev.next = self.next
+        if self.next:
+            self.next.prev = self.prev
+
 
 """Our doubly-linked list class. It holds references to
 the list's head and tail nodes."""
+
+
 class DoublyLinkedList:
-  def __init__(self, node=None):
-    self.head = node
-    self.tail = node
-    self.length = 1 if node is not None else 0
+    def __init__(self, node=None):
+        self.head = node
+        self.tail = node
+        self.length = 1 if node is not None else 0
+
+    def __len__(self):
+        return self.length
+
+    """
+    For debugging
+    """
+
+    def __repr__(self):
+        node = self.head
+        string = ""
+        if node is None:
+            return "[ ]"
+        while node.next is not None:
+            string += f"{node.value}, "
+            node = node.next
+        return f"[ {string}{node.value} ]"
 
-  def __len__(self):
-    return self.length
+    def add_to_head(self, value):
+        if self.length == 0:
+            node = ListNode(value)
+            self.head = node
+            self.tail = node
+        else:
+            self.head.insert_before(value)
+            self.head = self.head.prev
+        self.length += 1
 
-  def add_to_head(self, value):
-    pass
+    def remove_from_head(self):
+        if self.length == 0:
+            return None
+        elif self.length == 1:
+            node = self.head
+            self.tail = None
+            self.head = None
+            self.length -= 1
+            return node.value
+        else:
+            node = self.head
+            self.head.delete()
+            self.head = self.head.next
+            self.length -= 1
+            return node.value
 
-  def remove_from_head(self):
-    pass
+    def add_to_tail(self, value):
+        if self.length == 0:
+            node = ListNode(value)
+            self.tail = node
+            self.head = node
+        else:
+            self.tail.insert_after(value)
+            self.tail = self.tail.next
+        self.length += 1
 
-  def add_to_tail(self, value):
-    pass
+    def remove_from_tail(self):
+        if self.length == 0:
+            return None
+        elif self.length == 1:
+            node = self.tail
+            self.head = None
+            self.tail = None
+            self.length -= 1
+            return node.value
+        else:
+            node = self.tail
+            self.tail = self.tail.prev
```

Comment:
i think you are forgetting to clear tail.next after you assigned self.tail to be self.tail.prev. self.tail.prev would have a next reference that you do not want to have if you make that to be your new self.tail